tests/proxy/pn_counter/pn_counter_test.go

```diff
@@ -0,0 +1,210 @@
+	res, err := counter.Get()
+	AssertEqualf(t, err, res, int64(0), "PNCounter.Destroy failed")
+}
+
+func TestPNCounter_Get(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	counter.AddAndGet(delta)
+	currentValue, err := counter.Get()
+	AssertEqualf(t, err, currentValue, delta, "PNCounter.Get failed")
+}
+
+func TestPNCounter_GetAndAdd(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	previousValue, err := counter.GetAndAdd(delta)
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndAdd failed")
+}
+
+func TestPNCounter_AddAndGet(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	updatedValue, err := counter.AddAndGet(delta)
+	AssertEqualf(t, err, updatedValue, delta, "PNCounter.AddAndGet failed")
+}
+
+func TestPNCounter_GetAndSubtract(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	previousValue, err := counter.GetAndSubtract(delta)
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndSubtract failed")
+}
+
+func TestPNCounter_SubtractAndGet(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	updatedValue, err := counter.SubtractAndGet(delta)
+	AssertEqualf(t, err, updatedValue, -delta, "PNCounter.SubtractAndGet failed")
+}
+
+func TestPNCounter_DecrementAndGet(t *testing.T) {
+	defer destroyAndCreate()
+	updatedValue, err := counter.DecrementAndGet()
+	AssertEqualf(t, err, updatedValue, int64(-1), "PNCounter.DecrementAndGet failed")
+}
+
+func TestPNCounter_IncrementAndGet(t *testing.T) {
+	defer destroyAndCreate()
+	updatedValue, err := counter.IncrementAndGet()
+	AssertEqualf(t, err, updatedValue, int64(1), "PNCounter.IncrementAndGet failed")
+}
+
+func TestPNCounter_GetAndDecrement(t *testing.T) {
+	defer destroyAndCreate()
+	previousValue, err := counter.GetAndDecrement()
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndDecrement failed")
+}
+
+func TestPNCounter_GetAndIncrement(t *testing.T) {
+	defer destroyAndCreate()
+	previousValue, err := counter.GetAndIncrement()
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndIncrement failed")
+}
+
+func TestPNCounter_ManyAdd(t *testing.T) {
+	defer destroyAndCreate()
+	var wg sync.WaitGroup
+	var delta = 1000
+	wg.Add(delta)
+	for i := 0; i < delta; i++ {
+		go func() {
+			counter.IncrementAndGet()
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	currentValue, err := counter.Get()
+	AssertEqualf(t, err, currentValue, int64(delta), "PNCounter has race condition")
+
+}
+
+func TestPNCounter_HazelcastNoDataMemberInClusterError(t *testing.T) {
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+	cluster, err = remoteController.CreateCluster("3.10", LiteMemberConfig)
+	remoteController.StartMember(cluster.ID)
+	client, _ = hazelcast.NewHazelcastClient()
+	counter, _ = client.GetPNCounter(counterName)
+	var delta int64 = 5
+	_, err = counter.AddAndGet(delta)
+	if _, ok := err.(*core.HazelcastNoDataMemberInClusterError); !ok {
```

Comment:
Can you add a `assertExpectError(t *Testing, err Error, msg string)` assertion to `util` and use that for this kind of checks.

Reply:
It should be like this:
 `AssertExpectedError(t, err, reflect.TypeOf((*core.HazelcastNoDataMemberInClusterError)(nil)).Elem(),
		"PNCounter.AddAndGet should return HazelcastNoDataMemberInClusterError")`

I think using an extra method does not make it easy or nice-looking.

Reply:
Can you pass `*core.HazelcastNoDataMemberInClusterError` directly instead of `reflect.TypeOf((*core.HazelcastNoDataMemberInClusterError)(nil)).Elem()`? Since `reflect.TypeOf` can take it as an argument, I am tempted to think that you can pass that type to another function and do the refecltion within you function. But I am not familiar with reflection in Golang so I will take your word for it.

I can agree that it does not look nicer that way but I still see some value in it.
- Having the assert method clearly shows that it is the intention of this test to ensure that the error is of that type.
- Less error prone to write since you delegate `if` branching here. 
- If you are fine with a more general error message such as `Expected $x error, but got $y`, you can avoid writing error description every time. I think general error message is good because all the information about the error is still easily reachable. Specific error message does not provide extra value.
- Easier to read.

Reply:
````
func TestUnreadRuneAtEOF(t *testing.T) {
	// UnreadRune/ReadRune should error at EOF (was a bug; used to panic)
	r := NewReader(strings.NewReader("x"))
	r.ReadRune()
	r.ReadRune()
	r.UnreadRune()
	_, _, err := r.ReadRune()
	if err == nil {
		t.Error("expected error at EOF")
	} else if err != io.EOF {
		t.Error("expected EOF; got", err)
	}
}
````
It is a built in `Go` library test. Tests are written by using `if` statement.

Reply:
This is a minor point. I am not strongly for one side.